api/src/services/authorization.js

```diff
@@ -344,10 +400,62 @@ const isSensitive = function(userCtx, subject, submitter, allowedSubmitter) {
   return !allowedSubmitter;
 };
 
+/**
+ * Returns whether a doc should be visible, depending on configured replication_depth.report_depth.
+ * We're receiving all relevant replication keys to check whether at least one is of valid depth.
+ * In cases of reports with `needs_signoff`, the replication key of valid depth might be at facility level, but that
+ * key would be marked as sensitive.
+ * @param {Object} authorizationContext
+ * @param {Object} authorizationContext.subjectsDepth
+ * @param {number} authorizationContext.subjectsDepth.subject - relative depth of every contact the user can see
+ * @param {Object} authorizationContext.reportDepth - the maximum allowed depth for replicated reports
+ * @param {Array[]} replicationKeys - array of pairs of subject + value, emitted by docs_by_replication_keys view
+ * @returns {boolean}
+ */
+const isAllowedDepth = (authorizationContext, replicationKeys) => {
+  if (!usesReportDepth(authorizationContext)) {
+    // no depth limitation
+    return true;
+  }
+
+  const docType = replicationKeys[0][1].type;
+  if (docType !== 'data_record') {
+    // allow everything that's not a data_record through (f.e. targets)
+    return true;
+  }
+
+  return replicationKeys.some(replicationKey => {
+    const [ subject, { submitter } ] = replicationKey;
+    if (submitter === authorizationContext.userCtx.contact_id) {
+      // current user is the submitter
+      return true;
+    }
+
+    return authorizationContext.subjectsDepth[subject] <= authorizationContext.reportDepth;
+  });
+};
+
+const groupViewResultsById = (authorizationContext, viewResults) => {
+  const viewResultsById = {};
+  if (!usesReportDepth(authorizationContext)) {
+    return viewResultsById;
+  }
+
+  viewResults.rows.forEach(row => {
+    if (!viewResultsById[row.id]) {
+      viewResultsById[row.id] = [];
+    }
+    viewResultsById[row.id].push([row.key, row.value]);
+  });
+
+  return viewResultsById;
```

Comment:
Can this be just

```js
return _.groupBy(viewResults.rows, 'id');
```

The result is a little different because it's an array of objects rather than arrays but I think that'll make the code easier to read anyway, ie: referencing by property name rather than array index.

Reply:
This format is the same that results when using `view-map-utils` to get view results for a doc. 
I'm also using `allowedDepth` in `allowedDoc` function where I have that exact result available, so I matched the "inputs". 

Do you think it'd be worth to change the output of `view-map-utils` from [key, value] to { key, value } ?